### PR TITLE
Add Spectrum acceleration for MiniMax H3

### PR DESCRIPTION
**Pull request by xmarre**
## Summary

Adds a standalone Spectrum-style acceleration custom node for ComfyUI's native MiniMax H3 audio-video model.

- wraps only cloned MiniMax H3 model patchers
- caches compact post-transformer target audio/video features on CPU
- forecasts selected solver steps with Chebyshev ridge history weights
- preserves native current-step output heads, reconstruction, sigma mapping, and return structure
- keeps unsupported samplers and multi-GPU parallel sampling on the native path
- leaves ComfyUI-Spectrum-Proper and its FLUX integration untouched

## Implementation

The acceleration boundary is the packed hidden state after the final H3 transformer block and before the native FinalLayer. Actual steps execute the native H3 forward path and observe the final block output through a call-local replacement. Forecast steps skip transformer blocks and invoke the native output path with exact current-step audio/video timestep embeddings.

Runtime state is clone-local and transactional. Failed or ambiguous branch correspondence aborts the forecast attempt and re-runs the entire predict-noise transaction as an actual step.

## Validation

- `python -m ruff check .`
- `python tests/smoke_forecaster.py`
- `COMFYUI_PATH=<attached ComfyUI source> PYTHONPATH=<attached ComfyUI source> python -m pytest -q repo`
- Result: `37 passed`
- Forced-actual fixture matches native H3 video/audio output exactly
- Forecast fixture executes zero H3 transformer blocks
- Release archive and SHA-256 checksum are generated only after the tested `main` commit succeeds

Native-equivalence development target: ComfyUI commit `e377e263049f9338b4d12a3dd417b36ae62948ff`.

## Known validation limits

A real MiniMax H3 checkpoint has not yet validated:

- text-to-video/audio generation and reference modes
- decoded video and audio quality
- audiovisual synchronization
- long-duration CPU memory behavior
- VRAM/RSS peaks
- end-to-end wall-clock improvement

The README labels presets and performance expectations as provisional until those measurements exist.

## Summary by CodeRabbit

* **New Features**
  * Added a ComfyUI node for applying MiniMax H3 spectrum sampling.
  * Added configurable blending, forecasting, history, scheduling, memory, and debugging options.
  * Added adaptive sampling with supported-sampler detection and automatic fallback to standard execution.
  * Added native MiniMax H3 model validation and safe model cloning.

* **Documentation**
  * Added installation, configuration, workflow, performance, fallback, and implementation documentation.

* **Tests**
  * Added comprehensive coverage for forecasting, layouts, runtime scheduling, native equivalence, and wrapper behavior.

* **Chores**
  * Added packaging metadata, ignore rules, and automated test workflows.

### CODE REVIEW

**Comment by xmarre** [comment]
@coderabbitai review

File: comfyui_spectrum_h3/minimax_h3.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+    required = (
+        "blocks",
+        "final_layer",
+        "hidden_size",
+        "patch_size",
+        "latents_dim",
+        "audio_latents_dim",
+        "sigma_shift_video",
+        "sigma_shift_audio",
+    )
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/minimax_h3.py:49]: @coderabbitai Addressed in `f979661` with a necessary correction to the proposed contract. Native H3 creates `adaln_t_table` only for curve checkpoints and `time_embedder` only for non-curve checkpoints, so requiring both would reject valid native models. Validation now requires boolean `use_adaln_curves` plus the corresponding conditional attribute; the fake and tests cover both valid modes and both missing-attribute failures.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    steps:
+      - name: Check out custom node
+        uses: actions/checkout@v4
+
+      - name: Check out reviewed ComfyUI source
+        uses: actions/checkout@v4
+        with:
+          repository: comfyanonymous/ComfyUI
+          ref: e377e263049f9338b4d12a3dd417b36ae62948ff
+          path: ComfyUI
```
[the remark this answers is not in the document]
Comment (xmarre) [reply .github/workflows/tests.yml:26]: @coderabbitai Fixed in `f979661`: both `actions/checkout@v4` steps now set `persist-credentials: false`, while the native ComfyUI checkout keeps its existing repository, ref, and path.

File: comfyui_spectrum_h3/config.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            raise ValueError("window_size must be finite and >= 1")
+        if not math.isfinite(self.flex_window) or self.flex_window < 0.0:
+            raise ValueError("flex_window must be finite and >= 0")
+        if isinstance(self.warmup_steps, bool) or self.warmup_steps < 0:
+            raise ValueError("warmup_steps must be an integer >= 0")
+        if isinstance(self.tail_actual_steps, bool) or self.tail_actual_steps < 0:
+            raise ValueError("tail_actual_steps must be an integer >= 0")
+        if isinstance(self.max_history, bool) or self.max_history < self.min_fit_points:
+            raise ValueError(
+                f"max_history must be an integer >= {self.min_fit_points} for degree {self.degree}"
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/config.py]: @coderabbitai Fixed in `f979661`: all four count fields now reject booleans and non-`int` values before `min_fit_points` is used. Added parameterized coverage for fractional `degree`, `warmup_steps`, `tail_actual_steps`, and `max_history` values.

File: comfyui_spectrum_h3/minimax_h3.py [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+    inner = executor.class_obj
+    if not is_native_minimax_h3(inner):
+        return executor(
+            x,
+            timestep,
+            context,
+            options,
+            minimax_payload=minimax_payload,
+            **kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/minimax_h3.py:390]: @coderabbitai Fixed in `f979661`, including the missing second half of the transaction. The non-native branch now calls `fallback_current_step`, and `finalize_step` accepts a disabled fallback with zero H3 calls as a completed native passthrough. The suggested fallback call alone would still have reached the existing zero-call `RuntimeError`. A regression test executes and finalizes this exact path.

File: comfyui_spectrum_h3/runtime.py [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+        if self._run is None:
+            raise RuntimeError("Spectrum H3 runtime is outside a sampling run")
+        value_tensor = torch.as_tensor(timestep).detach().to(device="cpu", dtype=torch.float64).reshape(-1)
+        if value_tensor.numel() == 0 or not bool(torch.isfinite(value_tensor).all().item()):
+            raise RuntimeError("current solver timestep is empty or nonfinite")
+        if not bool(torch.allclose(value_tensor, value_tensor[0].expand_as(value_tensor))):
+            raise RuntimeError("current predict_noise call contains multiple solver timesteps")
+        value = float(value_tensor[0].item())
+        coordinate = 2.0 * (value - self._run.sigma_min) / (self._run.sigma_max - self._run.sigma_min) - 1.0
+        return float(max(-1.0, min(1.0, coordinate)))
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/runtime.py]: @coderabbitai Fixed in `f979661`: `coordinate_for_timestep` returns neutral coordinate `0.0` for a non-finite or non-positive sigma span. Added coverage for an invalid equal-sigma schedule reaching the disabled native path.

File: comfyui_spectrum_h3/sampling.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            noise,
+            latent_image,
+            sampler,
+            sigmas,
+            denoise_mask,
+            callback,
+            disable_pbar,
+            seed,
+            latent_shapes=latent_shapes,
+        )
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/sampling.py:108]: @coderabbitai Addressed in `f979661`: the README now explicitly requires the H3/packed-latent sampler APIs at ComfyUI commit `e377e263049f9338b4d12a3dd417b36ae62948ff`, calls out `outer_sample(..., latent_shapes=...)`, marks older revisions unsupported, and marks later revisions unverified.

File: tests/test_native_equivalence.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        SpectrumH3Config(degree=1, max_history=4, force_actual=True, warmup_steps=0, tail_actual_steps=0)
+    )
+    run_id = runtime.start_run(torch.tensor([1.0, 0.5, 0.0]), "sample_euler", supported_sampler=True)
+    wrapped, _ = _wrapped_call(model, runtime, 1.0, 500.0, x, context, payload)
+    assert native_options == before
+    assert isinstance(wrapped, list) and len(wrapped) == 2
+    for native_part, wrapped_part in zip(native, wrapped):
+        assert native_part.shape == wrapped_part.shape
+        assert native_part.dtype == wrapped_part.dtype
+        torch.testing.assert_close(native_part, wrapped_part, rtol=0.0, atol=0.0)
```
[the remark this answers is not in the document]
Comment (xmarre) [reply tests/test_native_equivalence.py]: @coderabbitai Fixed in `f979661`: the native/wrapped equivalence loop now uses `zip(..., strict=True)`, so output-part count mismatches fail before shape, dtype, and exact-value comparisons.

File: tests/smoke_forecaster.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    prediction = forecaster.predict(0.5, blend_weight=0.5)
+    assert prediction.shape == (2, 3, 4)
+    assert prediction.dtype == torch.float16
+    assert torch.isfinite(prediction).all()
+    assert forecaster.persistent_tensor_bytes < forecaster.history_tensor_bytes + 4096
+    print("Spectrum H3 forecaster smoke test passed")
+
+
+if __name__ == "__main__":
+    main()
```
[the remark this answers is not in the document]
Comment (xmarre) [reply tests/smoke_forecaster.py:29]: @coderabbitai Fixed in `f979661`: CI now has a dedicated `python tests/smoke_forecaster.py` step before the native pytest suite. The smoke script remains directly executable and its assertions are unchanged.

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+Repository = "https://github.com/xmarre/ComfyUI-Spectrum-MiniMax-H3"
+Issues = "https://github.com/xmarre/ComfyUI-Spectrum-MiniMax-H3/issues"
+
+[tool.comfy]
+PublisherId = "xmarre"
+DisplayName = "Spectrum MiniMax H3"
+Icon = ""
+
+[tool.setuptools.packages.find]
+include = ["comfyui_spectrum_h3*"]
```
[the remark this answers is not in the document]
Comment (xmarre) [reply pyproject.toml:34]: @coderabbitai Fixed in `f979661` by moving the functional ComfyUI node entry point into `comfyui_spectrum_h3/nodes.py` and exporting its mappings from the installed package; the root `nodes.py` remains a source-checkout compatibility shim. The built wheel was inspected and installed in isolation, and `comfyui_spectrum_h3.NODE_CLASS_MAPPINGS` imports successfully from it.

**Comment by xmarre** [comment]
@coderabbitai Nitpick reply — strict history pairing: fixed in `f979661`. Both `zip(weight_scalars, self._history)` and the direct-equivalence fixture’s `zip(coords, features)` now use `strict=True`, making their equal-length invariants executable.

**Comment by xmarre** [comment]
@coderabbitai Nitpick reply — pinned host history: reviewed; no code change. Each archive is a large, long-lived full-feature tensor. Pinning every retained entry would consume page-locked system memory and add a full archival copy, while the current per-chunk copy followed by same-stream accumulation does not establish useful overlap without GPU benchmarks. Keeping bounded pageable CPU history has the lower memory and stability risk; this optimization should be reconsidered only with real H3 transfer/VRAM/RSS measurements.

**Comment by xmarre** [comment]
@coderabbitai Nitpick reply — adaLN fake attributes: addressed in `f979661` with the native conditional contract. ComfyUI defines `adaln_t_table` for curve checkpoints and `time_embedder` for non-curve checkpoints, never both on the reviewed constructor path. The fake now models both modes, and tests verify rejection when the mode-specific attribute is absent.

**Comment by xmarre** [comment]
@coderabbitai Nitpick reply — adaptive-window bound: fixed in `f979661`. Post-warmup growth remains rounded and is clamped to `max(window_size, max_history)`. A regression test uses an oversized `flex_window` and verifies the cap.

**Comment by xmarre** [comment]
@coderabbitai Nitpick reply — incompatible history updates: fixed in `f979661`. `forecaster.update` is wrapped for `ValueError`; a shape/dtype change disables forecasting, clears incompatible history, and lets actual-step accounting and reset complete. A regression test changes actual-feature dtype and confirms native progress continues.

**Comment by xmarre** [comment]
@coderabbitai Nitpick reply — future replacement contract: reviewed critically. I did not add source/bytecode fingerprinting or a synthetic callback probe. Callback output cannot be proven without executing the native block, and source fingerprints would reject compatible ComfyUI changes. The README now states the exact tested API baseline and that later revisions are unverified; apply-time validation checks the native class and mode-specific attributes, while actual execution validates the replacement’s `{"img": tensor}` result and packed-layout shape with explicit contract errors.

File: tests/smoke_forecaster.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (xmarre) [reply tests/smoke_forecaster.py:29]: @coderabbitai Follow-up in `702f197`: the first CI run exposed that direct script execution uses `tests/` as its import root on a clean runner. The smoke step now sets `PYTHONPATH` to `${{ github.workspace }}`. I reproduced the clean-path failure with the repository root absent and the passing invocation with it present; the smoke script itself remains unchanged.
